src/kernel/state/vatState.js

```diff
@@ -0,0 +1,327 @@
+import harden from '@agoric/harden';
+import Nat from '@agoric/nat';
+import { insist } from '../insist';
+
+export default function makeVatState() {
+  // per-vat translation tables
+
+  // kernelSlotToVatSlot is an object with four properties:
+  //    exports, devices, promises, resolvers.
+  // vatSlotToKernelSlot has imports, deviceImports, promises, resolvers
+  const state = {
+    kernelSlotToVatSlot: {
+      exports: new Map(),
+      devices: new Map(),
+      promises: new Map(),
+      resolvers: new Map(),
+    },
+    vatSlotToKernelSlot: {
+      imports: new Map(),
+      deviceImports: new Map(),
+      promises: new Map(),
+      resolvers: new Map(),
+    },
+
+    // make these IDs start at different values to detect errors
+    // better
+    nextIDs: {
+      import: 10,
+      promise: 20,
+      resolver: 30,
+      deviceImport: 40,
+    },
+    transcript: [],
+  };
+
+  const allowedVatSlotTypes = [
+    'export',
+    'import',
+    'deviceImport',
+    'promise',
+    'resolver',
+  ];
+
+  const allowedKernelSlotTypes = ['export', 'device', 'promise', 'resolver'];
+
+  function insistVatSlot(slot) {
+    const properties = Object.getOwnPropertyNames(slot);
+    insist(
+      properties.length === 2,
+      `wrong number of properties for a vatSlot ${JSON.stringify(slot)}`,
+    );
+    Nat(slot.id);
+    insist(
+      allowedVatSlotTypes.includes(slot.type),
+      `unknown slot.type in '${JSON.stringify(slot)}'`,
+    );
+  }
+
+  function insistKernelSlot(slot) {
+    const properties = Object.getOwnPropertyNames(slot);
+    insist(
+      properties.length === 3 ||
+        (properties.length === 2 &&
+          (slot.type === 'promise' || slot.type === 'resolver')),
+      `wrong number of properties for a kernelSlot ${JSON.stringify(slot)}`,
+    ); // a kernel slot has a vatID property, unless it is a promise or a resolver
+    Nat(slot.id);
+    insist(
+      allowedKernelSlotTypes.includes(slot.type),
+      `unknown slot.type in '${JSON.stringify(slot)}'`,
+    );
+  }
+
+  function getKernelSlotTypedMapAndKey(kernelSlot) {
+    insistKernelSlot(kernelSlot);
+
+    const { type, id } = kernelSlot;
+    let typedKernelToVatMap;
+    let kernelSlotKey;
+
+    switch (type) {
+      case 'promise': {
+        typedKernelToVatMap = state.kernelSlotToVatSlot.promises;
+        kernelSlotKey = `${type}-${id}`;
```

Comment:
Let's think about how to make this more concise. We don't need the `type-` in each key because we're using separate maps for each type. And there's a lot of repetition here, which makes the code hard to read in a single glance.. if we can fit it on a single screen without being too clever, I think it'll be easier to read. Maybe something like:

```
const tables = state.kernelSlotToVatSlot;
switch (type) {
 case 'promise':
  return [ tables.promises, id];
 case 'resolver':
  return [ tables.resolvers, id];
 case 'export':
  return [ tables.exports, `${kernelSlot.vatID}-${id}` ];
 ... etc
```

and change the caller to expect a two-element list instead of an object.

Reply:
Sure, I can do that. I was a little hesitant to optimize this, because I think anything specifically related to JavaScript maps will be going away in the next step when we switch to using the `merk` merkle tree.

Reply:
yeah, good thought, although I'm guessing we'll have separate subtrees for each Map, so the code will have a similar structure (instead of returning a `[Map, key]` pair, we'll be returning a `[merkle subtree node, key]` pair.